src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/SourceItemHandler.cs

```diff
@@ -98,50 +90,65 @@ public void Handle(CommandLineArguments added, CommandLineArguments removed)
                 {
                     string fullPath = _project.MakeRooted(filePath);
 
-                    AddSourceFile(fullPath, treeState);
+                    AddSourceFile(fullPath, context, treeState);
                 }
 
                 foreach (KeyValuePair<string, string> filePaths in diff.RenamedItems)
                 {
                     string beforeFullPath = _project.MakeRooted(filePaths.Key);
                     string afterFullPath = _project.MakeRooted(filePaths.Value);
 
-                    RemoveSourceFile(beforeFullPath);
-                    AddSourceFile(afterFullPath, treeState);
+                    RemoveSourceFile(beforeFullPath, context);
+                    AddSourceFile(afterFullPath, context, treeState);
                 }
 
                 foreach (string filePath in diff.ChangedItems)
                 {   // We add and then remove ChangedItems to handle Linked metadata changes
 
                     string fullPath = _project.MakeRooted(filePath);
 
-                    RemoveSourceFile(fullPath);
-                    AddSourceFile(fullPath);
+                    RemoveSourceFile(fullPath, context);
+                    AddSourceFile(fullPath, context);
                 }
             }
         }
 
-        private void RemoveSourceFile(string fullPath)
+        public override Task OnContextReleasedAsync(IWorkspaceProjectContext context)
+        {
+            _sourceFilesByContext.Remove(context);
+            return Task.CompletedTask;
+        }
+
+        private void RemoveSourceFile(string fullPath, IWorkspaceProjectContext context)
         {
-            if (_sourceFiles.Remove(fullPath))
+            HashSet<string> sourceFiles;
+            if (_sourceFilesByContext.TryGetValue(context, out sourceFiles) && sourceFiles.Remove(fullPath))
             {
-                _context.RemoveSourceFile(fullPath);
+                context.RemoveSourceFile(fullPath);
             }
         }
 
-        private void AddSourceFile(string fullPath, IProjectTreeServiceState state = null)
+        private void AddSourceFile(string fullPath, IWorkspaceProjectContext context, IProjectTreeServiceState state = null)
         {
-            if (!_sourceFiles.Contains(fullPath))
+            HashSet<string> sourceFiles;
```

Comment:
That is what should probably be done in the long run, but right now we are working with singleton instances of handlers. I'll file a work item to track this improvement.